### DISCUSSION TITLE
Can't override search with a 3rd party search?

### DISCUSSION THREAD

**Discussion by A-Baji** [squidfunk/mkdocs-material#4077] [Q&A] [answered]
I am trying to implement a 3rd party search since the native mkdocs-material search is not suitable for my project. I haven't chosen which search to use but right now I am testing with [AddSearch](https://www.addsearch.com/). It crawls the provided domain and returns a key that is used here:
`<script src="https://cdn.addsearch.com/v4/addsearch-ui.min.js?key=####"></script>`

I'd hoped that it would be as simple as overriding the search using this script like so:
```
{% extends "base.html" %}

{% block config %}
  {{ super() }}
  <script>
    var __search = {
      worker: "https://cdn.addsearch.com/v4/addsearch-ui.min.js?key=####"
    }
  </script>
{% endblock %}
```
But doing this causes the search bar to disappear altogether. Am I doing something wrong? Or is what I want to do not possible?

**Comment by squidfunk** [comment, marked as answer]
The `worker` must point to a valid web worker. I'd suspect that you'd need to add the script as a separate search script, or somehow integrate querying their API into the web worker. Possibly not as easy as just including their script.

**Reply by A-Baji** [reply]
> add the script as a separate search script

Would you happen to have any examples of this that I could take a look at?

> integrate querying their API into the web worker

Something with [this](https://www.npmjs.com/package/addsearch-js-client)?

Sorry my javascript is quite weak.

**Reply by squidfunk** [reply]
I'm afraid I don't have any examples. I've also never heard of AddSearch. [Customizing search](https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-search/#custom-search) is not trivial, you'll probably need some solid frontend expertise for that. In general, you need to replace the webworker with an implementation that mimics the same API, as described in the documentation.

Maybe other users can share their customization here!

**Reply by UltraInstinct05** [reply]
> since the native mkdocs-material search is not suitable for my project.

Maybe you could provide where exactly the default search fails for your use case ? Could be something that could be improved in `mkdocs-material`. The default search has been really solid for me, one of the best among various documentation systems I have tried out.

**Reply by A-Baji** [reply]
> Maybe you could provide where exactly the default search fails for your use case ?

I am trying to create a documentation hub for various projects. One of the requirements for this is that all child documentation must include `mike` versioning. The thing about `mike` is that it was implemented to work specifically with `gh-pages`. The only workaround for this that I could figure out was to pull directly from each projects' `gh-pages` build into the parent docs directory, like this:
![image](https://user-images.githubusercontent.com/61991709/177370111-c461733e-50c0-4e47-af30-5241758f49e0.png)
Then to access a child repo's docs, I just relatively link it in the parent `index.md`.

This all works perfectly. The only issue is that searching in the parent doc does not search any of the children docs, since the search index does not index the children. As I'm typing this, I'm realizing that a 3rd party search is probably not the best option. Would merging the child search indexes into the parent index while altering the locations accordingly work?

Perhaps this could be a feature of `mkdocs-material`? Where users can include in the config paths to search indexes that they want to be merged. Or is that too niche of a function?

**Reply by squidfunk** [reply]
> Would merging the child search indexes into the parent index while altering the locations accordingly work?

That would be my recommendation. The `search_index.json` format is very straight forward, so you'd essentially only need to merge all nested `search_index.json` into one and prepend the path or URL to each entry. Probably 10-20 lines of code.

**Reply by squidfunk** [reply]
> Perhaps this could be a feature of mkdocs-material? Where users can include in the config paths to search indexes that they want to be merged. Or is that too niche of a function?

It's rather niche. Also, this is a perfect use case for a third-party MkDocs plugin. It's probably already supported by one of the multi-project plugins.

**Reply by A-Baji** [reply]
> It's probably already supported by one of the multi-project plugins.

I will look into this before trying to do it myself. Thanks for the help.
